### PR TITLE
SLCORE-182 add support for connected mode in language server

### PR DESCRIPTION
**Pull request by janos-ss**

### CODE REVIEW

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java [hunk 1]
```diff
@@ -220,34 +250,143 @@ void error(String message, Throwable t) {
...
+      debug("Could not parse the value of " + CONNECTED_MODE_SERVERS_PROP + ": " + connectedModeServers);
+    }
+  }
+
+  private static boolean isBlank(@Nullable String s) {
+    return s == null || s.isEmpty();
+  }
+
+  private void handleUpdateServerStorageCommand(@Nullable List<Object> arguments) {
+    // TODO only update servers that have changed
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWMr4Qgbau6S9RpGg66y&open=AWMr4Qgbau6S9RpGg66y)

[ISSUE_KEY]: AWMr4Qgbau6S9RpGg66y

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java [hunk 2]
```diff
@@ -174,12 +196,16 @@ private UserSettings(Map<String, Object> params) {
       this.disableTelemetry = (Boolean) params.getOrDefault(DISABLE_TELEMETRY, false);
     }
 
-    private static Map getAnalyzerProperties(Map<String, Object> params) {
-      Map analyzerProperties = (Map) params.get(ANALYZER_PROPERTIES);
-      if (analyzerProperties == null) {
-        analyzerProperties = Collections.emptyMap();
+    private Map<String, String> getAnalyzerProperties(Map<String, Object> params) {
+      try {
+        Map<String, String> analyzerProperties = (Map<String, String>) params.get(ANALYZER_PROPERTIES);
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Rename "analyzerProperties" which hides the field declared at line 186. ([squid:HiddenFieldCheck](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AHiddenFieldCheck))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWMr4Qgbau6S9RpGg66x&open=AWMr4Qgbau6S9RpGg66x)

[ISSUE_KEY]: AWMr4Qgbau6S9RpGg66x

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java [hunk 3]
```diff
@@ -550,21 +740,32 @@ public WorkspaceService getWorkspaceService() {
...
           String severity = ruleDetails.getSeverity();
           client.openRuleDescription(RuleDescription.of(ruleKey, ruleName, htmlDescription, type, severity));
         }
         break;
+      case SONARLINT_UPDATE_SERVER_STORAGE_COMMAND:
+        List<Object> list = args == null ? null : args.stream().map(SonarLintLanguageServer::parseToMap).collect(Collectors.toList());
+        handleUpdateServerStorageCommand(list);
+        break;
+      case SONARLINT_UPDATE_PROJECT_BINDING_COMMAND:
+        Map<String, Object> map = args == null || args.size() < 1 ? null : parseToMap(args.get(0));
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Use isEmpty() to check whether the collection is empty or not. ([squid:S1155](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1155))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWMr4Qgbau6S9RpGg66z&open=AWMr4Qgbau6S9RpGg66z)

[ISSUE_KEY]: AWMr4Qgbau6S9RpGg66z

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java (hunk 1 above)
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWMr4Qgbau6S9RpGg66y&open=AWMr4Qgbau6S9RpGg66y)

[ISSUE_KEY]: AWMr4Qgbau6S9RpGg66y

Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWMr4Qgbau6S9RpGg66y&open=AWMr4Qgbau6S9RpGg66y)

[ISSUE_KEY]: AWMr4Qgbau6S9RpGg66y

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java [hunk 4]
```diff
@@ -153,12 +177,12 @@ public SonarLintLanguageServer(InputStream inputStream, OutputStream outputStrea
...
   }
 
-  public static SonarLintLanguageServer bySocket(int port, Collection<URL> analyzers) throws IOException {
+  static SonarLintLanguageServer bySocket(int port, Collection<URL> analyzers) throws IOException {
     Socket socket = new Socket("localhost", port);
     return new SonarLintLanguageServer(socket.getInputStream(), socket.getOutputStream(), analyzers);
   }
 
-  private static class UserSettings {
+  private class UserSettings {
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Make this a "static" inner class. ([squid:S2694](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2694))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWM3TWeR_vaxgEW1HRnk&open=AWM3TWeR_vaxgEW1HRnk)

[ISSUE_KEY]: AWM3TWeR_vaxgEW1HRnk

Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Make this a "static" inner class. ([squid:S2694](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2694))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWM3TWeR_vaxgEW1HRnk&open=AWM3TWeR_vaxgEW1HRnk)

[ISSUE_KEY]: AWM3TWeR_vaxgEW1HRnk

File: language-server/src/main/java/org/sonarlint/languageserver/Logger.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+  public void info(String message) {
+
+  }
+
+  public void warn(String message) {
+
+  }
+
+  public void error(String message, Exception e) {
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/Logger.java]: Code Smell Code Smell: Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. ([squid:S1186](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1186))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWND8ndh6NykrOWnW1BT&open=AWND8ndh6NykrOWnW1BT)

[ISSUE_KEY]: AWND8ndh6NykrOWnW1BT

File: language-server/src/main/java/org/sonarlint/languageserver/ConnectedEngineFactory.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+import org.sonarsource.sonarlint.core.client.api.connected.ConnectedGlobalConfiguration;
+import org.sonarsource.sonarlint.core.client.api.connected.ConnectedSonarLintEngine;
+
+/**
+ * Create and start connected engines based on server info, configured with loggers and extra properties.
+ */
+class ConnectedEngineFactory {
+
+  private final LogOutput logOutput;
+  private final Logger logger;
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/ConnectedEngineFactory.java]: Code Smell Code Smell: Make the "logger" logger private static final and rename it to comply with the format "LOG(?:GER)?". ([squid:S1312](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1312))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWND8neb6NykrOWnW1BX&open=AWND8neb6NykrOWnW1BX)

[ISSUE_KEY]: AWND8neb6NykrOWnW1BX

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java [hunk 7]
```diff
@@ -286,6 +296,103 @@ private static TextDocumentSyncOptions getTextDocumentSyncOptions() {
...
+
+  private void updateServerStorage(ConnectedSonarLintEngine engine, ServerInfo serverInfo) {
+    String serverId = serverInfo.serverId;
+    debug("Updating global storage of server " + serverId + ", may take some time...");
+    ServerConfiguration serverConfig = getServerConfiguration(serverInfo);
+    engine.update(serverConfig, null);
+    debug("Successfully updated global storage of server " + serverId);
+  }
+
+  private ServerConfiguration getServerConfiguration(ServerInfo serverInfo) {
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Make "getServerConfiguration" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWND8neO6NykrOWnW1BW&open=AWND8neO6NykrOWnW1BW)

[ISSUE_KEY]: AWND8neO6NykrOWnW1BW

File: language-server/src/main/java/org/sonarlint/languageserver/ServerInfoCache.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+import javax.annotation.Nullable;
+
+import static org.apache.commons.lang.StringUtils.isBlank;
+
+/**
+ * Cache of server connection details. Used for updating server and module storage.
+ */
+class ServerInfoCache {
+
+  private final Logger logger;
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/ServerInfoCache.java]: Code Smell Code Smell: Make the "logger" logger private static final and rename it to comply with the format "LOG(?:GER)?". ([squid:S1312](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1312))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWND8neu6NykrOWnW1BZ&open=AWND8neu6NykrOWnW1BZ)

[ISSUE_KEY]: AWND8neu6NykrOWnW1BZ

File: language-server/src/main/java/org/sonarlint/languageserver/Logger.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+  public void warn(MessageType messageType) {
+    warn(messageType.message);
+  }
+
+  public void info(String message) {
+
+  }
+
+  public void warn(String message) {
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/Logger.java]: Code Smell Code Smell: Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. ([squid:S1186](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1186))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWND8ndh6NykrOWnW1BS&open=AWND8ndh6NykrOWnW1BS)

[ISSUE_KEY]: AWND8ndh6NykrOWnW1BS

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java [hunk 10]
```diff
@@ -151,14 +176,23 @@ public SonarLintLanguageServer(InputStream inputStream, OutputStream outputStrea
...
+    BiFunction<LanguageClientLogOutput, Logger, EngineCache> factory = (logOutput, logger) -> {
+      StandaloneEngineFactory standaloneEngineFactory = new StandaloneEngineFactory(analyzers, logOutput, logger);
+      ConnectedEngineFactory connectedEngineFactory = new ConnectedEngineFactory(logOutput, logger);
+      return new DefaultEngineCache(standaloneEngineFactory, connectedEngineFactory);
+    };
+    return new SonarLintLanguageServer(socket.getInputStream(), socket.getOutputStream(), factory);
   }
 
-  private static class UserSettings {
+  private class UserSettings {
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Make this a "static" inner class. ([squid:S2694](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2694))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWM3TWeR_vaxgEW1HRnk&open=AWM3TWeR_vaxgEW1HRnk)

[ISSUE_KEY]: AWM3TWeR_vaxgEW1HRnk

File: language-server/src/main/java/org/sonarlint/languageserver/StandaloneEngineFactory.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+import org.sonarsource.sonarlint.core.client.api.standalone.StandaloneGlobalConfiguration;
+import org.sonarsource.sonarlint.core.client.api.standalone.StandaloneSonarLintEngine;
+
+/**
+ * Create and start standalone engines, configured with loggers and extra properties.
+ */
+class StandaloneEngineFactory {
+  private final Collection<URL> analyzers;
+  private final LogOutput logOutput;
+  private final Logger logger;
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/StandaloneEngineFactory.java]: Code Smell Code Smell: Make the "logger" logger private static final and rename it to comply with the format "LOG(?:GER)?". ([squid:S1312](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1312))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWND8neo6NykrOWnW1BY&open=AWND8neo6NykrOWnW1BY)

[ISSUE_KEY]: AWND8neo6NykrOWnW1BY

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java [hunk 12]
```diff
@@ -104,43 +104,68 @@
...
+  static final String SONARLINT_UPDATE_PROJECT_BINDING_COMMAND = "SonarLint.UpdateProjectBinding";
+  private static final List<String> SONARLINT_COMMANDS = Arrays.asList(
+    SONARLINT_OPEN_RULE_DESCRIPTION_COMMAND,
+    SONARLINT_UPDATE_SERVER_STORAGE_COMMAND,
+    SONARLINT_UPDATE_PROJECT_BINDING_COMMAND);
 
   private final SonarLintLanguageClient client;
   private final Future<?> backgroundProcess;
   private final LanguageClientLogOutput logOutput;
+  private final Logger logger;
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Make the "logger" logger private static final and rename it to comply with the format "LOG(?:GER)?". ([squid:S1312](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1312))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWND8neO6NykrOWnW1BV&open=AWND8neO6NykrOWnW1BV)

[ISSUE_KEY]: AWND8neO6NykrOWnW1BV

File: language-server/src/main/java/org/sonarlint/languageserver/Logger.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    MessageType(String message) {
+      this.message = message;
+    }
+  }
+
+  public void warn(MessageType messageType) {
+    warn(messageType.message);
+  }
+
+  public void info(String message) {
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/Logger.java]: Code Smell Code Smell: Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. ([squid:S1186](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1186))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWND8ndh6NykrOWnW1BR&open=AWND8ndh6NykrOWnW1BR)

[ISSUE_KEY]: AWND8ndh6NykrOWnW1BR

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java [hunk 14]
```diff
@@ -151,14 +169,27 @@ public SonarLintLanguageServer(InputStream inputStream, OutputStream outputStrea
...
     Socket socket = new Socket("localhost", port);
-    return new SonarLintLanguageServer(socket.getInputStream(), socket.getOutputStream(), analyzers);
+
+    BiFunction<LanguageClientLogOutput, ClientLogger, EngineCache> engineCacheFactory = (logOutput, logger) -> {
+      StandaloneEngineFactory standaloneEngineFactory = new StandaloneEngineFactory(analyzers, logOutput, logger);
+      ConnectedEngineFactory connectedEngineFactory = new ConnectedEngineFactory(logOutput, logger);
+      return new DefaultEngineCache(standaloneEngineFactory, connectedEngineFactory);
+    };
+
+    Function<SonarLintLanguageClient, ClientLogger> loggerFactory = client -> new DefaultClientLogger(client);
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Replace this lambda with a method reference. ([squid:S1612](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1612))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWNEPxp56NykrOWnW1Uv&open=AWNEPxp56NykrOWnW1Uv)

[ISSUE_KEY]: AWNEPxp56NykrOWnW1Uv

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java [hunk 15]
```diff
@@ -151,14 +169,27 @@ public SonarLintLanguageServer(InputStream inputStream, OutputStream outputStrea
...
+
+    return new SonarLintLanguageServer(socket.getInputStream(), socket.getOutputStream(), engineCacheFactory, loggerFactory);
   }
 
-  private static class UserSettings {
+  private class UserSettings {
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Make this a "static" inner class. ([squid:S2694](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2694))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWM3TWeR_vaxgEW1HRnk&open=AWM3TWeR_vaxgEW1HRnk)

[ISSUE_KEY]: AWM3TWeR_vaxgEW1HRnk

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java [hunk 16]
```diff
@@ -174,34 +205,15 @@ private UserSettings(Map<String, Object> params) {
       this.disableTelemetry = (Boolean) params.getOrDefault(DISABLE_TELEMETRY, false);
     }
 
-    private static Map getAnalyzerProperties(Map<String, Object> params) {
-      Map analyzerProperties = (Map) params.get(ANALYZER_PROPERTIES);
-      if (analyzerProperties == null) {
-        analyzerProperties = Collections.emptyMap();
+    private Map<String, String> getAnalyzerProperties(Map<String, Object> params) {
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Make "getAnalyzerProperties" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWNEQd1A6NykrOWnW1WD&open=AWNEQd1A6NykrOWnW1WD)

[ISSUE_KEY]: AWNEQd1A6NykrOWnW1WD

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java [hunk 17]
```diff
@@ -570,23 +748,36 @@ public WorkspaceService getWorkspaceService() {
...
+        if (args == null) {
+          break;
+        }
         if (args.size() != 1) {
-          warn("Expecting 1 argument");
+          logger.warn("Expecting 1 argument");
         } else {
           String ruleKey = parseToString(args.get(0));
-          RuleDetails ruleDetails = engine.getRuleDetails(ruleKey);
+          // TODO use the correct engine (currently only Atom reaches this code, and it doesn't have connected mode, yet)
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWNFGjsklRUrfRWxk7UR&open=AWNFGjsklRUrfRWxk7UR)

[ISSUE_KEY]: AWNFGjsklRUrfRWxk7UR

File: language-server/src/main/java/org/sonarlint/languageserver/ServerInfoCache.java [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+import static org.apache.commons.lang.StringUtils.isBlank;
+
+/**
+ * Cache of server connection details. Used for updating server and module storage.
+ */
+class ServerInfoCache {
+
+  private final ClientLogger logger;
+
+  private Map<String, ServerInfo> cache = new HashMap<>();
```
Comment (dbmeneses) [inline language-server/src/main/java/org/sonarlint/languageserver/ServerInfoCache.java]: can be final

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java (hunk 17 above)
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWNFGjsklRUrfRWxk7UR&open=AWNFGjsklRUrfRWxk7UR)

[ISSUE_KEY]: AWNFGjsklRUrfRWxk7UR

**Pull request merged**